metasrv/tests/it/grpc/metasrv_grpc_watch.rs

```diff
@@ -0,0 +1,250 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::thread;
+use std::time::Duration;
+
+use common_base::tokio;
+use common_base::tokio::sync::mpsc;
+use common_base::tokio::sync::mpsc::UnboundedReceiver;
+use common_base::tokio::sync::mpsc::UnboundedSender;
+use common_meta_api::KVApi;
+use common_meta_grpc::MetaGrpcClient;
+use common_meta_types::protobuf::event::SeqV;
+use common_meta_types::protobuf::watch_request::FilterType;
+use common_meta_types::protobuf::Event;
+use common_meta_types::protobuf::WatchRequest;
+use common_meta_types::MatchSeq;
+use common_meta_types::Operation;
+use common_meta_types::UpsertKVAction;
+
+use crate::init_meta_ut;
+
+async fn watcher_client_main(
+    addr: String,
+    watch: WatchRequest,
+    start_tx: UnboundedSender<()>,
+    mut watch_events: Vec<Event>,
+) -> anyhow::Result<()> {
+    let client = MetaGrpcClient::try_create(addr.as_str(), "root", "xxx", None, None).await?;
+    let mut grpc_client = client.make_client().await?;
+
+    let request = tonic::Request::new(watch);
+
+    let mut client_stream = grpc_client.watch(request).await?.into_inner();
+
+    // notify client stream has started
+    let _ = start_tx.send(());
+
+    loop {
+        tokio::select! {
+            resp = client_stream.message() => {
+                if let Ok(Some(resp)) = resp {
+                    if let Some(event) = resp.event {
+                        assert!(!watch_events.is_empty());
+
+                        assert_eq!(watch_events.get(0), Some(&event));
+                        watch_events.remove(0);
+
+                        if watch_events.is_empty() {
+                            // notify has recv all the notify events
+                            let _ = start_tx.send(());
+                        }
+                    }
+
+                }
+            },
+        }
+    }
+}
+
+fn wait_notify(start_rx: &mut UnboundedReceiver<()>, wait_ms: i32) {
+    let mut cnt = 0;
+    while start_rx.try_recv().is_err() {
+        thread::sleep(Duration::from_millis(100));
+        cnt += 100;
+        if cnt >= wait_ms {
+            panic!("wait out of time");
+        }
+    }
+}
+
+async fn test_watch_main(
+    addr: String,
+    watch: WatchRequest,
+    events: Vec<Event>,
+    updates: Vec<UpsertKVAction>,
+) -> anyhow::Result<()> {
+    let client = MetaGrpcClient::try_create(addr.as_str(), "root", "xxx", None, None).await?;
+
+    let (start_tx, mut start_rx) = mpsc::unbounded_channel::<()>();
+
+    let h = tokio::spawn(watcher_client_main(addr.clone(), watch, start_tx, events));
+
+    // wait for client stream start up
+    wait_notify(&mut start_rx, 2000);
+
+    // update some kv
+    for update in updates.iter() {
+        let _ = client.upsert_kv(update.clone()).await;
+    }
```

Comment:
I mean, put these 3 lines in to a task, then you can just remove `watcher_client_main` along with the channels for communication: just move the body of `watcher_client_main` here.

Reply:
`client.upsert_kv` must wait for watcher has been created, so use the channel to notify the watcher has been created, only after that the `update_kv` has meaning in this case.

Reply:
That's not a problem. You can just create the watcher, then spawn the task, then drain the response stream to check the events.